ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/deduplication/CommandDeduplicationBase.scala

```diff
@@ -477,39 +524,121 @@ private[testtool] abstract class CommandDeduplicationBase(
-        )
+        )(additionalErrorAssertions)
       )
-  }
 
-  protected def submitRequestAndAssertAsyncDeduplication(ledger: ParticipantTestContext)(
+  protected def submitRequestAndAssertAsyncDeduplication(
+      ledger: ParticipantTestContext
+  )(
       request: SubmitRequest,
       parties: Party*
+  )(
+      additionalCompletionAssertions: Completion => Unit = _ => ()
   )(implicit ec: ExecutionContext): Future[OffsetWithCompletion] =
-    submitRequestAndAssertCompletionStatus(
+    submitRequestAndAssertCompletion(
       ledger
-    )(request, Code.ALREADY_EXISTS, parties: _*)
+    )(
+      request,
+      parties: _*
+    ) { completion =>
+      assertCompletionStatus(request, completion, Code.ALREADY_EXISTS)
+      additionalCompletionAssertions(completion)
+    }
+
+  private def assertCompletionStatus(
+      request: SubmitRequest,
+      completion: Completion,
+      statusCode: Code,
+  ): Unit =
+    assert(
+      completion.getStatus.code == statusCode.value(),
+      s"""Expecting completion with status code $statusCode but completion has status ${completion.status}.
+         |Request: $request
+         |Completion: $completion
+         |Metadata: ${extractErrorInfoMetadata(
+        GrpcStatus.toJavaProto(completion.getStatus)
+      )}""".stripMargin,
+    )
 
-  private def submitRequestAndAssertCompletionStatus(
+  private def assertErrorOffsetAndSubmissionId(
+      optCommands: Option[Commands],
+      expectedCompletionOffset: String,
+  )(
+      t: Throwable
+  ): Unit = t match {
+    case exception @ Exception =>
+      val metadata = extractErrorInfoMetadata(exception)
+      Option(metadata.get("completion_offset")) match {
+        case Some(value) =>
+          assertEquals(
+            "completion offset mismatch",
+            value,
+            expectedCompletionOffset,
+          )
+        case _ =>
+      }
+      Option(metadata.get("existing_submission_id")) match {
+        case someSubmissionId @ Some(_) =>
+          assertEquals(
+            "submission ID mismatch",
+            someSubmissionId,
+            optCommands.map(_.submissionId),
+          )
+        case _ =>
+      }
+    case _ => ()
+  }
+
+  private def assertCompletionOffsetAndSubmissionId(
+      optCommands: Option[Commands],
+      expectedCompletionOffset: String,
+  )(
+      completion: Completion
+  ): Unit = {
+    val metadata = extractErrorInfoMetadata(
+      GrpcStatus.toJavaProto(completion.getStatus)
+    )
+    Option(metadata.get("completion_offset")) match {
+      case Some(value) =>
+        assertEquals(
+          "completion offset mismatch",
+          value,
+          expectedCompletionOffset,
+        )
+      case _ =>
+    }
+    Option(metadata.get("existing_submission_id")) match {
+      case someSubmissionId @ Some(_) =>
+        assertEquals(
+          "submission ID mismatch",
+          someSubmissionId,
+          optCommands.map(_.submissionId),
```

Comment:
As we use only the submission id from the commands, can't we accept only the submission id in the params so that we don't pass down the entire command?

Reply:
Yes, I am doing several refactorings and that's one of them, upcoming.

Reply:
9e4b6c0